botocore/handlers.py

```diff
@@ -148,11 +148,20 @@ def _calculate_md5_from_file(fileobj):
     return md5.digest()
 
 
-def conditionally_calculate_md5(params, **kwargs):
-    """Only add a Content-MD5 when not using sigv4"""
-    signer = kwargs['request_signer']
-    if signer.signature_version not in ['v4', 's3v4'] and MD5_AVAILABLE:
-        calculate_md5(params, **kwargs)
+def conditionally_calculate_md5(params, context, request_signer, **kwargs):
+    """Only add a Content-MD5 if the system supports it."""
+    if not MD5_AVAILABLE:
+        return
+
+    s3_config = context.get('client_config').s3
+    if s3_config is None:
+        s3_config = {}
+    use_sha256 = s3_config.get('sha256_sign_s3v4_payload', False)
+    if request_signer.signature_version == 's3v4' and \
```

Comment:
Questions for everyone: if the users opt-in to sha256_sign_s3v4_payload, should we also do MD5 calculation? Based on the docstring you have it is not clear that you turn off MD5 signing.